Simple gates can be executed (#7432)
PennyLaneAI/pennylane: doc/releases/changelog-dev.md

<h3>New features since last release</h3>

> * A new function :func:`~.from_qasm3` is provided that allows very simple circuits composed of series of basic
>   gate applications to be converted into callables that may be loaded into QNodes and executed. Most QASM 3.0 features
>   are not yet supported. [(#7432)](https://github.com/PennyLaneAI/pennylane/pull/7432)

Reviewer: ```suggestion
* A new function called :func:`~.from_qasm3` has been added, which converts OpenQASM3 circuits 
  into quantum functions that may be loaded into QNodes and executed. 
  [(#7432)](https://github.com/PennyLaneAI/pennylane/pull/7432)
```

  ```python
  import pennylane as qml

>   from pennylane import device, wires
>   from pennylane.io import from_qasm3

Reviewer: This is perfectly fine! But we usually just have `import pennylane as qml` and call everything from that :) 

```suggestion
```

Reviewer 2: > But we usually just have import pennylane as qml and call everything from that :)

Import police here 🚨 👮🏼   - Just want to reiterate that this should only be done if you are a user. 

Internally developed code shouldn't import from top-level. So in this case, since you are demonstrating an example of user code, this suggestion is correct.

>   execute_qasm, wires = from_qasm3("qubit q0; qubit q1; ry(0.2) q0; rx(1.0) q1; pow(2) @ x q0;")

Reviewer: ```suggestion
  execute_qasm, wires = qml.from_qasm3("qubit q0; qubit q1; ry(0.2) q0; rx(1.0) q1; pow(2) @ x q0;")
```

  dev = device("default.qubit", wires=[w for w in wires])
  
  @qml.qnode(dev)
  def my_circuit():
      execute_qasm()
      return qml.expval(qml.Z(0))
  ```

  ```pycon
  >>> print(qml.draw(my_circuit)())
  0: ──RY(0.20)──X²─┤  <Z>
  1: ──RX(1.00)─────┤  

>   ```

Reviewer: ````suggestion
  ```

  Some gates and operations in OpenQASM3 programs are not currently supported. For more details, 
  please consult the documentation for :func:`~.from_qasm3`.
````

<h3>New features since last release</h3>

* A new function :func:`~.from_qasm3` is provided that allows very simple circuits composed of series of basic
  gate applications to be converted into callables that may be loaded into QNodes and executed. Most QASM 3.0 features
  are not yet supported. [(#7432)](https://github.com/PennyLaneAI/pennylane/pull/7432)

  ```python
  import pennylane as qml
  from pennylane import device
  from pennylane.io import from_qasm3

  dev = device("default.qubit", wires=2)
  
  @qml.qnode(dev)
  def my_circuit(y, p):
      execute_qasm, _ = from_qasm3(f"""
          qubit q0;
          qubit q1;
          float theta = 0.5;
          x q0;
          cx q0, q1;
          rx(theta) q0;
          ry({y}) q0;
          inv @ rx(theta) q0;
          pow({p}) @ x q0;
          ctrl @ x q1, q0;
          """
      )
      execute_qasm()
      return qml.expval(qml.Z(0))
  

>   my_circuit(y=0.2, p=2)

Reviewer 3: Maybe replace this with `print(qml.draw(my_circuit)(0.2, 2))` and actually display the output.

Author: Sure, good suggestion!

Author: Done!

<h3>New features since last release</h3>

* A new function :func:`~.from_qasm3` is provided that allows very simple circuits composed of series of basic
  gate applications to be converted into callables that may be loaded into QNodes and executed. Most QASM 3.0 features

>   are not yet supported. [(#7432)](https://github.com/PennyLaneAI/pennylane/pull/7432)

Reviewer 3: It's fine for now, but just letting you know that in all your follow up PRs you would modify this message, and add each of them to the same changelog entry.

Author: Are you talking about how there are some line returns in my message?

Reviewer 3: I mean in this changelog entry you say that "most qasm 3.0 features are not yet supported", but in your follow up PRs you'll be adding support to them. So when you do that, you wouldn't add another changelog entry. Instead you would modify this changelog entry to not say that, and add the link to your follow up PRs here.

Author: oh, sure, makes sense.

  ```python
  import pennylane as qml
  from pennylane import device, wires
  from pennylane.io import from_qasm3

>   dev = device("default.qubit", wires=[wires.Wires('q0'), wires.Wires('q1'), 0])

Author: Actually I don't think that there is any change necessary to `from_qasm3` since it does not do anything with `Wires`. This is just how I provided wires to the device in this example in the changelog.

>   dev = device("default.qubit", wires=[wires.Wires('q0'), wires.Wires('q1'), 0])

Author: I will also change to docstring for `from_qasm3` though, maybe that's what you meant.

>   dev = device("default.qubit", wires=[wires.Wires('q0'), wires.Wires('q1'), 0])

Author: Done!

  
  @qml.qnode(dev)
  def my_circuit(y, p):
      execute_qasm, _ = from_qasm3(f"""
          qubit q0;
          qubit q1;
          float theta = 0.5;
          x q0;
          cx q0, q1;
          rx(theta) q0;
          ry({y}) q0;
          inv @ rx(theta) q0;
          pow({p}) @ x q0;
          ctrl @ x q1, q0;
          """

>       )

Reviewer 3: I thought the idea was that you use the wires returned from the `from_qasm3` to construct the device?

Author: I think that is still an intended use case, yes. In this case though I wanted to show how you can also call it within the QNode in order to paramterize the QASM with the QNode's parameters.

Reviewer 3: Oh I see what you mean, although I feel that this could be confusing. @albi3ro what do you think?

Reviewer 4: I think for the examples we should use the qasm wires to construct the device and measure one of them.

Author: Thanks for the feedback, I will update the example accordingly.

      execute_qasm()
      return qml.expval(qml.Z(0))
  

>   print(qml.draw(my_circuit)(0.2, 2))

Author: sure!

>   print(qml.draw(my_circuit)(0.2, 2))

Author: I have moved this to the `pycon` section. :)
